In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
from seaborn import regplot
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [2]:
df = pd.read_csv("./Weight-S2017-18.csv")
print(df.shape)
df

(100, 3)


height sex  weight
0     1.69   M   67.89
1     1.32   F   32.21
2     1.77   M   82.71
3     1.53   F   51.50
4     1.80   M   85.40
5     1.63   M   71.18
6     1.47   F   40.88
7     1.59   F   48.45
8     1.77   M   78.99
9     1.54   F   45.71
10    1.58   F   46.13
11    1.75   M   71.58
12    1.81   M   82.96
13    1.70   M   69.97
14    1.78   M   81.88
15    1.45   F   40.02
16    1.50   F   45.68
17    1.86   M   87.43
18    1.50   F   47.65
19    1.36   F   35.74
20    1.62   F   51.78
21    1.60   F   52.41
22    1.53   F   43.91
23    1.63   M   59.54
24    1.67   M   60.39
25    1.82   M   81.48
26    1.78   M   82.78
27    1.51   F   39.81
28    1.58   M   60.57
29    1.72   M   77.13
..     ...  ..     ...
70    1.57   M   62.58
71    1.43   F   43.40
72    1.49   F   50.18
73    1.59   M   62.02
74    1.62   F   55.05
75    1.71   M   72.28
76    1.74   F   62.40
77    1.54   M   58.79
78    1.73   M   77.11
79    1.55   M   65.37
80    1.66   M   70.81
81    1.56   F   55.73
82    1.62   M   65.83
83    1.69   M   69.99
84    1.39   F   36.41
85    1.43   F   41.54
86    1.49   F   45.40
87    1.43   F   43.64
88    1.68   M   69.40
89    1.78   M   74.71
90    1.32   F   35.20
91    1.82   M   85.86
92    1.69   M   69.57
93    1.44   F   42.79
94    1.46   M   49.83
95    1.53   F   38.53
96    1.41   F   39.85
97    1.52   F   47.74
98    1.79   M   79.62
99    1.39   F   37.76

[100 rows x 3 columns]

In [3]:
features = df.drop('weight',axis=1)
features = pd.get_dummies(features)
features.head(3)

height  sex_F  sex_M
0    1.69      0      1
1    1.32      1      0
2    1.77      0      1

In [4]:
target = df.weight
target.head(3)

0    67.89
1    32.21
2    82.71
Name: weight, dtype: float64

In [5]:
features = (features - features.mean())/features.std()
lr = LinearRegression()
lr.fit(features,target)
R2 = lr.score(features,target)
print('R-squared =',round(R2,3))

coef = pd.DataFrame()
coef['feature'] = features.columns
coef['coef'] = lr.coef_.round(2)
coef.sort_values('coef',ascending=False)

R-squared = 0.955


feature   coef
0  height  11.82
2   sex_M   2.60
1   sex_F  -2.60

## (i) *Justifications provided for method used*

#### I knew right off the bat that I wanted to use a Regression Model, since the output of my prediction is a number, not a model (model = use a classifier). From the plot we are given, there appears to be a rough, linear relationship between height and weight. From this, I decided a Linear Regression algorithm would be best. This would keep things simple, yet accurate.

#### I one-hot encode the qualitative feature in the data set: 'sex'. This is I can fit the quantitative feature 'height' with the qualitative feature of 'sex'.

#### There is a need to standardize the variables in this case. There is collinearity between the independent variables. The independent variables are 'height' and 'sex'. One could use sex to predict height; changing sex will have an effect on height, as women are generally shorter than men.

## (ii) *Simplicity and interpretability of method used and interpretations provided for results*

#### As one can see, a person's height has a positive coefficient, meaning height has a positive correlation with weight. Same with being a male, but to a lesser degree. We can see that women have negative correlation with weight, meaning, keeping height constant, a woman is more likely to weigh less than a man.

#### Let's get specific. See that height's coefficient is 11.82. This means, give a one unit change in height increases weight by 11.82 units; a one meter increase in height results in a 11.82 kg weight gain, on average. If a person is male, their weight usually increases by 2.6 kg, and if they're a women, their weight usually decreases by 2.6 kg

## (iii) *Accuracy of predictions for weight*

In [6]:
df = pd.read_csv("./Weight_predict-S2017-18.csv")
df.head(3)

height sex  weight
0    1.48   F       0
1    1.80   M       0
2    1.75   M       0

In [7]:
features = df.drop('weight',axis=1)
features = pd.get_dummies(features)
features.head(3)
features = (features - features.mean())/features.std()

In [8]:
predictions = lr.predict(features)

In [9]:
df['weight'] = predictions
df

height sex     weight
0     1.48   F  46.907505
1     1.80   M  81.726261
2     1.75   M  77.906290
3     1.70   M  74.086319
4     1.64   M  69.502354
5     1.54   F  51.491471
6     1.72   M  75.614308
7     1.48   F  46.907505
8     1.81   M  82.490255
9     1.45   F  44.615523
10    1.68   M  72.558331
11    1.76   M  78.670284
12    1.53   M  61.098418
13    1.49   F  47.671500
14    1.46   F  45.379517
15    1.74   M  77.142296
16    1.49   F  47.671500
17    1.52   F  49.963482
18    1.76   M  78.670284
19    1.37   F  38.503569
20    1.55   F  52.255465
21    1.56   F  53.019459
22    1.49   F  47.671500
23    1.56   F  53.019459
24    1.64   M  69.502354
25    1.79   M  80.962267
26    1.77   M  79.434279
27    1.48   F  46.907505
28    1.62   F  57.603424
29    1.49   M  58.042441
..     ...  ..        ...
70    1.43   F  43.087534
71    1.40   F  40.795552
72    1.53   F  50.727477
73    1.48   F  46.907505
74    1.50   F  48.435494
75    1.34   F  36.211587
76    1.67   M  71.794337
77    1.60   F  56.075436
78    1.94   M  92.422180
79    1.75   M  77.906290
80    1.66   M  71.030342
81    1.59   M  65.682383
82    1.38   F  39.267563
83    1.74   M  77.142296
84    1.67   M  71.794337
85    1.17   F  23.223685
86    1.78   M  80.198273
87    1.52   M  60.334423
88    1.47   F  46.143511
89    1.74   M  77.142296
90    1.72   M  75.614308
91    1.79   M  80.962267
92    1.56   F  53.019459
93    1.66   M  71.030342
94    1.75   M  77.906290
95    1.73   M  76.378302
96    1.40   F  40.795552
97    1.66   M  71.030342
98    1.65   M  70.266348
99    1.69   M  73.322325

[100 rows x 3 columns]

In [10]:
df.to_csv("Weight_rileyma.csv",index=False)
df.head(3)

height sex     weight
0    1.48   F  46.907505
1    1.80   M  81.726261
2    1.75   M  77.906290